controlplane/kubeadm/controllers/controller.go

```diff
@@ -580,3 +592,44 @@ func (r *KubeadmControlPlaneReconciler) adoptOwnedSecrets(ctx context.Context, k
 
 	return nil
 }
+
+func (r *KubeadmControlPlaneReconciler) remediateUnhealthy(ctx context.Context, controlPlane *internal.ControlPlane) error {
+	if !controlPlane.RemediationAllowed() || !controlPlane.HasUnhealthyMachine() {
+		return nil
+	}
+	if controlPlane.HasDeletingMachine() {
+		return &capierrors.RequeueAfterError{RequeueAfter: deleteRequeueAfter}
+	}
+
+	// Get the workload cluster client.
+	workloadCluster, err := r.managementCluster.GetWorkloadCluster(ctx, util.ObjectKey(controlPlane.Cluster))
+	if err != nil {
+		return errors.Wrap(err, "failed to create workload cluster client")
+	}
+
+	etcdStatus, err := workloadCluster.EtcdStatus(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to fetch etcd status")
+	}
+
+	if etcdStatus.FailureTolerance() == 0 {
```

Comment:
I think this should be sophisticated a little bit, because we should consider what happens if we delete a specific machine instead of checking if it is possible to delete any machine.

e.g. 3CP, 1 unhealthy (kubelet not ready, but etcd healthy), but another machine healthy from a MHC standpoint, by with the etcd member unhealthy.
In this case, I can delete theoretically delete an etcd member, but if I delete the unhealthy machine I will delete one of the healthy etcd members and loose quorum

Reply:
Hmm, I'm not sure what we can detect about unhealthy etcds. We don't configure any alarms afaik and so the `status.Errors` always seems to be an empty list. I agree with you that this scenario is possible but I'm not sure how to detect it.